File: android/src/main/kotlin/africa/ejara/trustdart/coins/XLM.kt
import com.google.protobuf.ByteString
import wallet.core.java.AnySigner
import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet
import africa.ejara.trustdart.Coin
import africa.ejara.trustdart.utils.base64String
import africa.ejara.trustdart.utils.toLong
import wallet.core.jni.proto.Stellar
import wallet.core.jni.proto.Stellar.SigningOutput
import wallet.core.jni.StellarPassphrase


class XLM : Coin("XLM", CoinType.STELLAR) {

    enum class NetworkType(val passphrase: String) {
        MAINNET("Public Global Stellar Network ; September 2015"),
        TESTNET("Test SDF Network ; September 2015");
        companion object {
            fun fromString(network: String?): NetworkType {
                return when (network?.lowercase()) {
                    "testnet" -> TESTNET
                    else -> MAINNET // Default to mainnet
                }
            }
        }
    }

    override fun getPublicKey(path: String, mnemonic: String, passphrase: String): String? {
        val wallet = HDWallet(mnemonic, passphrase)
        return wallet.getKey(coinType, path).publicKeyEd25519.data().base64String()
    }

    override fun getRawPublicKey(path: String, mnemonic: String, passphrase: String): ByteArray? {
        val wallet = HDWallet(mnemonic, passphrase)
        return wallet.getKey(coinType, path).publicKeyEd25519.data()
    }

    override fun signTransaction(
        path: String,
        txData: Map<String, Any>,
        mnemonic: String,
        pass_phrase: String
    ): String? {
        val cmd = txData["cmd"] as String
        val networkType: NetworkType = {
            val network = txData["network"] as? String
            NetworkType.fromString(network)
        }()
        
        val secretKey = HDWallet(mnemonic, pass_phrase).getKey(coinType, path)
        val txHash: String?
        when (cmd) {
            "ChangeTrust" -> {
                val assetUsdt = Stellar.Asset.newBuilder()
                assetUsdt.apply {
                    issuer = txData["toAddress"] as String
                    alphanum4 = txData["assetCode"] as String
                }
                val operation = Stellar.OperationChangeTrust.newBuilder()
                operation.apply {
                    asset = assetUsdt.build()
                    validBefore =
                        txData["validBefore"]!!.toLong()  // till when the asset trust is valid (timestamp)
                }
                val signingInput = Stellar.SigningInput.newBuilder()
                signingInput.apply {
                    account = txData["ownerAddress"] as String
                    fee = txData["fee"] as Int
                    sequence = txData["sequence"]!!.toLong()
                    passphrase = networkType.passphrase
                    opChangeTrust = operation.build()
                    privateKey = ByteString.copyFrom(secretKey.data())
                }
                val output = AnySigner.sign(signingInput.build(), coinType, SigningOutput.parser())
                txHash = output.signature
            }
            "Payment" -> {
                val stellarAsset = Stellar.Asset.newBuilder()
                if (txData["asset"] != null && txData["issuer"] != null) {
                    stellarAsset.apply {
                        issuer = txData["issuer"] as String
                        alphanum4 = txData["asset"] as String
                    }
                }
                val operation = Stellar.OperationPayment.newBuilder()
                operation.apply {
                    destination = txData["toAddress"] as String
                    amount = txData["amount"]!!.toLong()
                    if (txData["asset"] != null) {
                        asset = stellarAsset.build()
                    }
                }
                val signingInput = Stellar.SigningInput.newBuilder()
                signingInput.apply {
                    account = txData["ownerAddress"] as String
                    fee = txData["fee"] as Int
                    sequence = txData["sequence"]!!.toLong()
                    passphrase = networkType.passphrase
                    opPayment = operation.build()
                    privateKey = ByteString.copyFrom(secretKey.data())
                    if (txData["memo"] != null) {
                        memoId = Stellar.MemoId.newBuilder().setId(txData["memo"]!!.toLong())
                            .build()
                    }
                }
                val output = AnySigner.sign(signingInput.build(), coinType, SigningOutput.parser())
                txHash = output.signature

            }

            else -> txHash = null
        }
        return txHash
    }

}

File: ios/Classes/coins/XLM.swift
/*
 XLM
 */
import WalletCore

class XLM: Coin  {
    enum NetworkType: String {
        case mainnet
        case testnet
        
        var passphrase: String {
            switch self {
            case .mainnet:
                return "Public Global Stellar Network ; September 2015"
            case .testnet:
                return "Test SDF Network ; September 2015"
            }
        }
    }
    
    init() {
        super.init(name: "XLM", coinType: .stellar)
    }
    
    override func getPublicKey(path: String, mnemonic: String, passphrase: String) -> String? {
        let wallet = HDWallet(mnemonic: mnemonic, passphrase: passphrase)
        return wallet!.getKey(coin: self.coinType, derivationPath: path).getPublicKeyEd25519().data.base64EncodedString()
    }
    
    override func getRawPublicKey(path: String, mnemonic: String, passphrase: String) -> FlutterStandardTypedData {
        let wallet = HDWallet(mnemonic: mnemonic, passphrase: passphrase)
        return FlutterStandardTypedData(bytes: wallet!.getKey(coin: self.coinType, derivationPath: path).getPublicKeyEd25519().data)
    }
    
    override func signTransaction(path: String, txData: [String : Any], mnemonic: String, passphrase: String) -> String? {
        let privateKey = HDWallet(mnemonic: mnemonic, passphrase: passphrase)?.getKey(coin: self.coinType, derivationPath: path)
        let cmd = txData["cmd"] as! String
        var txHash: String?
        
        let networkType: NetworkType = {
            if let network = txData["network"], let type = NetworkType(rawValue: network as! String) {
                return type
            } else {
                return .mainnet // Default to mainnet if network is not provided or invalid
            }
        }()
        
        switch(cmd){
        case "ChangeTrust":
            let asset = StellarAsset.with {
                $0.issuer = txData["toAddress"] as! String
                $0.alphanum4 = txData["assetCode"] as! String
            }
            
            let operation = StellarOperationChangeTrust.with {
                $0.asset = asset
                $0.validBefore = txData["validBefore"] as! Int64
            }
            
            let signingInput = StellarSigningInput.with {
                $0.account = txData["ownerAddress"] as! String
                $0.fee = txData["fee"] as! Int32
                $0.sequence = txData["sequence"] as! Int64
                $0.passphrase = networkType.passphrase
                $0.opChangeTrust = operation
                $0.privateKey = privateKey!.data
                if (txData["memo"] != nil) {
                    $0.memoID = StellarMemoId.with {
                        $0.id = txData["memo"] as! Int64
                    }
                }
            }
            
            let output: StellarSigningOutput = AnySigner.sign(input: signingInput, coin: self.coinType)
            txHash = output.signature
        case "Payment":
            var asset = StellarAsset()
            if let assetString = txData["asset"] as? String, let issuer = txData["issuer"] as? String {
                asset.alphanum4 = assetString
                asset.issuer = issuer
            }
            
            let operation = StellarOperationPayment.with {
                $0.destination = txData["toAddress"] as! String
                $0.amount = txData["amount"] as! Int64
                if (txData["asset"] != nil) {
                    $0.asset = asset
                }
            }
            
            let signingInput = StellarSigningInput.with {
                $0.account = txData["ownerAddress"] as! String
                $0.fee = txData["fee"] as! Int32
                $0.sequence = txData["sequence"] as! Int64
                $0.passphrase = networkType.passphrase
                $0.opPayment = operation
                $0.privateKey = privateKey!.data
                if (txData["memo"] != nil) {
                    $0.memoID = StellarMemoId.with {
                        $0.id = txData["memo"] as! Int64
                    }
                }
            }
            
            let output: StellarSigningOutput = AnySigner.sign(input: signingInput, coin: self.coinType)
            txHash = output.signature
            
        default:
            txHash = nil
        }
        
        return txHash
    }
}
